Bug 1607229  - Export build metadata to BigQuery

acmiyaguchi opened the pull request:
This adds a `rebuild-bigquery` command for creating a replica of the Django models in BigQuery. It also adds a hook to stream new entries in the models the table configured in the settings.

The most important piece here is the schema, which has been pieced together from the models and included schema for build artifacts. This is used when (re)creating the table.

[bug link](https://bugzilla.mozilla.org/show_bug.cgi?id=1607229) and #596

mozbhearsum commented on Dockerfile:
```diff
@@ -25,6 +25,16 @@ WORKDIR /app/
+RUN apt-get -y update && apt-get install -y curl gnupg
+
+# Install Cloud SDK: https://cloud.google.com/sdk/docs/quickstart-debian-ubuntu
+RUN echo "deb [signed-by=/usr/share/keyrings/cloud.google.gpg] http://packages.cloud.google.com/apt cloud-sdk main" | \
```
Are you sure this is needed? I don't see it being used by anything further down, and I don't think the Python module depends on it.

acmiyaguchi replied:
It might not be needed, but it's nice to have access to the command-line tools like `gsutil` and `bq` for management. I'll drop this commit after everything is tested, if it's not needed.

mozbhearsum replied:
Sounds good to me!

acmiyaguchi replied:
Ended up dropping this, because none of the management needs to be done from the CLI.

mozbhearsum commented on buildhub/main/management/commands/reindex-bigquery.py:
```diff
@@ -0,0 +1,44 @@
+        """See https://googleapis.dev/python/bigquery/latest/index.html."""
+
+        # TODO: Best place to put this variable? Implicit in GOOGLE_APPLICATION_CREDENTIALS
+        project_id = "buildhub2-bigquery-dev"
```
`settings.py` seems like it would the most appropriate place. It looks like this can be set explicitly through the `Client` API: https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client. I think it's fine to allow credentials to be configured through `GOOGLE_APPLICATION_CREDENTIALS`, but I'd prefer `project` to be explicit.

Most likely we'd want to do it for at least stage and prod (but maybe dev too?), so perhaps part of https://github.com/mozilla-releng/buildhub2/blob/ee348d79d646b689bd91f1f41dee65e857d16a4b/buildhub/settings.py#L195, and have that read from the environment?

acmiyaguchi replied:
Added the relevant settings, `BQ_PROJECT_ID`, `BQ_DATASET_ID`, and` BQ_TABLE_ID`.

mozbhearsum commented on buildhub/main/management/commands/reindex-bigquery.py:
```diff
@@ -0,0 +1,44 @@
+            errors = client.insert_rows(table, rows)
+            if errors:
+                # dump the the first error seen
+                raise Exception(errors[0])
```
If raising an Exception kills the command, it's probably best to log this and move on to the next entry instead; otherwise there's the potential for one un-importable entry to cause us to get stuck.

mozbhearsum commented on buildhub/main/management/commands/reindex-bigquery.py:
```diff
@@ -0,0 +1,44 @@
+        qualified_table_id = f"{project_id}.{dataset_id}.{table_id}"
+        table = client.get_table(qualified_table_id)
+
+        chunk_size = 10000
```
Can we make this configurable in `settings`/environment too? It'll be much easier if we find we need to tweak it in production.

acmiyaguchi commented:
Thank you for the helpful review, especially around the settings. Django is a bit mysterious to me.

> Just to make sure I understand: the intent here is that the command will more or less do a sync of postgres -> bigquery -- which means we'll get all of the old stuff imported? And the assumption is that we'll run this command every X minutes/hours to continually import new stuff too?

That's the idea, for the most part. I'm not entirely familiar with how this service is run. My understanding is that this command would act as backfill (with table management and a sync from postgres), while the ingestion piece that reads from SQS into the database would include a callback to send the message to BigQuery. 

Ingestion from the message queue:
https://github.com/mozilla-releng/buildhub2/blob/ee348d79d646b689bd91f1f41dee65e857d16a4b/buildhub/ingest/sqs.py#L139-L144

Callback to elasticsearch (and BigQuery, in this PR):
https://github.com/mozilla-releng/buildhub2/blob/ee348d79d646b689bd91f1f41dee65e857d16a4b/buildhub/main/models.py#L112-L114

mozbhearsum commented:
> > Just to make sure I understand: the intent here is that the command will more or less do a sync of postgres -> bigquery -- which means we'll get all of the old stuff imported? And the assumption is that we'll run this command every X minutes/hours to continually import new stuff too?
> 
> That's the idea, for the most part. I'm not entirely familiar with how this service is run. My understanding is that this command would act as backfill (with table management and a sync from postgres), while the ingestion piece that reads from SQS into the database would include a callback to send the message to BigQuery.

Ah, so we'd do prehaps a one-off (or two-off) backfill, and then most likely never run the command again?

acmiyaguchi commented:
That's the plan . Reindexing is probably not the best description for this command because BigQuery doesn't actually keep an index like a traditional database or a search service. It's more akin to Athena or Spark.

acmiyaguchi commented on schema.bigquery.json:
```diff
@@ -0,0 +1,140 @@
+[
```
This file was generated using the following commands

```bash
pip install pyyaml
cargo install jsonschema-transpiler

cat schema.yaml | \
python3 -c 'import json, sys, yaml; y=yaml.safe_load(sys.stdin.read()); print(json.dumps(y))' | \
jq '.schema' | \
jsonschema-transpiler --type bigquery --resolve panic > schema.bigquery.json
```

mozbhearsum replied:
Can you find an appropriate place in the code to add this comment?

acmiyaguchi replied:
Added a script `bin/transpile-schema-bigquery.sh`.

acmiyaguchi commented:
@mozbhearsum r? Do you know who would eventually run the `rebuild-bigquery` command, and configure the server to point to the project and dataset containing the up to date buildhub2 data?

acmiyaguchi commented on buildhub/main/bigquery.py:
```diff
@@ -0,0 +1,54 @@
+    dataset_id = settings.BQ_DATASET_ID
+    table_id = f"{project_id}.{dataset_id}.{settings.BQ_TABLE_ID}"
+
+    client = bigquery.Client(project=project_id)
```
This can also spawn an authentication error, which will cause an exception to be thrown if `BQ_ENABLED=True` and `GOOGLE_APPLICATION_CREDENTIAL` is not properly set up.

mozbhearsum commented:
> @mozbhearsum r? Do you know who would eventually run the rebuild-bigquery command, and configure the server to point to the project and dataset containing the up to date buildhub2 data?

This will be on CloudOps' plate, and we'll probably want to give it a try in dev or staging first once the code is committed.

mozbhearsum commented on buildhub/main/models.py:
```diff
@@ -111,6 +116,8 @@ def insert(cls, build, metadata=None, skip_validation=False, **kwargs):
             # If it returns something, it got created! Must inform Elasticsearch.
             send_to_elasticsearch(cls, build)
+            if settings.BQ_ENABLED:
+                send_to_bigquery(cls, build)
```
What will happen if inserting to postgres/elasticsearch succeeds, but bigquery fails? Will the request fail overall, or will we end up in an inconsistent state? If the latter, do we have a way to recover from that (eg: by resyncing).

acmiyaguchi replied:
We would end up in an inconsistent state. There are reasonable retry defaults with the BigQuery library, so if it does fail, it would mean there's a configuration error somewhere. For example, a change to the schema may cause data to be rejected from the BigQuery table if done in a backwards incompatible fashion.

There are a couple of different ways to recover from this state. One would be to create a new script that insert values starting from a particular timestamp, and then run a deduplicate query to bring the table up to date. Another would be to log/mark the builds that failed, and retry them at another time (or manually from a script).

mozbhearsum replied:
Fully agree - this is bound to happen at some point :)

mozbhearsum replied:
Thinking about this a bit more, it would be really, really nice if whatever process does the backfilling was quick enough that we could run it, eg: once an hour, to pick up anything that was missed. This would avoid the need for any manual intervention.

Do you have any idea how long a no-op run of this (where we check for existence of everything in BigQuery but don't need to insert anything) would take right now? If it takes too long this might not be possible without other changes.

acmiyaguchi replied:
I wrote up an estimate of the size of the primary key set that would need to be used to filter out data that hasn't been picked up. I think it would make for a great follow-up to this PR, since the task is well defined and builds upon the work here. The logic is not complicated: it would materialize the set of keys locally from the BigQuery table, use it find the set difference, and then insert the straggling rows back into BigQuery.

mozbhearsum commented on docker-compose.yml:
```diff
@@ -58,8 +58,11 @@ services:
     volumes:
       - $PWD:/app
+      # if credentials do not exist, mount an arbitrary file
+      - ${GOOGLE_APPLICATION_CREDENTIALS:-./README.rst}:/tmp/credentials
```
What's the need for this? Can we just default `GOOGLE_APPLICATION_CREDENTIALS` to nothing instead?

acmiyaguchi replied:
I've run into issues with CircleCI throwing an error if the path is too short (i.e. `./` or ` `), but this piece of code might be premature.

acmiyaguchi commented on tests/conftest.py:
```diff
@@ -95,6 +101,50 @@ def test_something(elasticsearch):
     build_index.delete(ignore=404)
 
 
+@runif_bigquery_testing_enabled
```
I can mock out the requests to the service. It's possible for this to be set up [with a real project](https://github.com/mozilla/gcp-ingestion/blob/e0bf43ffb3638e093653d74e019a4ba93e44c4c9/.circleci/config.yml#L180-L188) which would fall under the free tier. This would be preferable, since these tests verify the schema via insertion into a newly created table.

mozbhearsum replied:
Perhaps we can have both? Some simpler tests to check what we can (which can always be run locally, even if you're not connected to the internet or have any credentials), and some more complete tests we run in CI?

acmiyaguchi replied:
Yep, I wrote a few tests for checking that the client is being hit (or not) after insertion into the main model store. I think using the actual service for verifying correctness is useful, but it does require a bit of configuration.

wlach commented on tests/utils.py:
```diff
@@ -0,0 +1,29 @@
+    )
+    return pytest.mark.skipif(
+        not bigquery_testing_enabled, reason="requires valid GCP credentials"
+    )(func)
```
With missioncontrol (the old version) I actually mocked out these types of calls in the test so some degree of testing could be performed locally. I suspect this is probably a better approach though.

acmiyaguchi commented on .env-dist:
```diff
@@ -19,6 +19,13 @@ AWS_ACCESS_KEY_ID=
+DJANGO_BQ_ENABLED=False
+DJANGO_BQ_PROJECT_ID=
+#DJANGO_BQ_DATASET_ID=buildhub2
+#DJANGO_BQ_TABLE_ID=builds
```
The `DJANGO_` prefix is how django maps environment variables into the settings object, so this is indeed necessary for actual deployment. All of the variables are being used within the PR.

acmiyaguchi replied:
https://django-configurations.readthedocs.io/en/0.4-0/values/#default-behavior for reference.

acmiyaguchi commented on buildhub/main/models.py:
```diff
@@ -173,3 +180,9 @@ def bulk_insert(
+@receiver(post_save, sender=Build)
+def send_to_bigquery(sender, instance, **kwargs):
+    doc = instance.to_dict()
+    insert_build(doc)
```
This seems to only catch `elasticsearch.exceptions.TransportError`. The `google-cloud-bigquery` library also has [a built-in retry mechanism](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.retry.DEFAULT_RETRY.html#google.cloud.bigquery.retry.DEFAULT_RETRY), which I observed to be 3x10sec.

wlach replied:
Ah ok, perhaps there's no need for this then. Might be worth adding a comment mentioning this, for help in debugging scenarios

acmiyaguchi commented on buildhub/main/management/commands/rebuild-bigquery.py:
```diff
@@ -0,0 +1,100 @@
+
+        client = bigquery.Client(project=project_id)
+        try:
+            client.get_table(table_id)
```
This script would probably not be useful if a table already exists. There isn't a great mechanism for upserting (replacing rows) because BigQuery is append-only by nature.

wlach replied:
How much data are we talking about? Assuming we're talking data in the 10s/100s of megabytes, doing a compare of what's in buildhub vs. in bigquery then appending any rows missing from the latter seems like it might be reasonable? Yes it would take a while, but at least it would be automated

acmiyaguchi replied:
We're looking at 1,727,065 rows * 16 bytes (per md5 hash), or about 30 mb to do the initial filtering. It would make this function a bit more complicated. I imagine the proper way to do this is to do something like:

```sql
WITH bigquery_hashes as (...),
SELECT *
FROM build
LEFT JOIN bigquery_hashes
USING (build_hash)
WHERE build_hash is NULL
```

and then iterating over the results of this query to insert into the table.  I'd prefer to leave as future work (i.e. out of scope of this PR), but this would be a good thing to have and schedule periodically.

mozbhearsum replied:
Leaving this for later sounds good to me.

acmiyaguchi commented:
I've addressed most of the review comments here, though there are some desirable features that haven't been written. This PR aims to enable an initial backfill into a BigQuery table from the production database, as well as ongoing insertions into that table.

It's possible that there is some loss of data during replication. In the comments, I suggest leaving this for another PR, since it should be much easier to deal with once the machinery is put into motion. It might make sense to add an option to update the schema, and another option to only update documents that do not currently exist.

The command as it exists now should be enough to start the initial import.

mozbhearsum commented on buildhub/main/apps.py:
```diff
@@ -34,3 +36,15 @@ def _check_s3_bucket_urls():
+        table_id = f"{project_id}.{dataset_id}.{settings.BQ_TABLE_ID}"
+        if settings.BQ_ENABLED:
+            client = bigquery.Client(project=project_id)
+            client.get_table(table_id)
```
Hm. When I tried to run locally, this check actually prevents Buildhub from starting up if the table doesn't exist. Maybe we should create the table on startup? We need to deploy the app before we can run the import command, so we do need to deal with this case.

And it looks like it actually prevents the initial import too :laughing::
```
root@b67429ef7a8b:/app# python manage.py rebuild-bigquery
Traceback (most recent call last):
  File "manage.py", line 15, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 357, in execute
    django.setup()
  File "/usr/local/lib/python3.7/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python3.7/site-packages/django/apps/registry.py", line 122, in populate
    app_config.ready()
  File "/app/buildhub/main/apps.py", line 20, in ready
    self._check_bigquery_table()
  File "/app/buildhub/main/apps.py", line 50, in _check_bigquery_table
    client.get_table(table_id)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 591, in get_table
    api_response = self._call_api(retry, method="GET", path=table_ref.path)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 487, in _call_api
    return call()
  File "/usr/local/lib/python3.7/site-packages/google/api_core/retry.py", line 286, in retry_wrapped_func
    on_error=on_error,
  File "/usr/local/lib/python3.7/site-packages/google/api_core/retry.py", line 184, in retry_target
    return target()
  File "/usr/local/lib/python3.7/site-packages/google/cloud/_http.py", line 421, in api_request
    raise exceptions.from_http_response(response)
google.api_core.exceptions.NotFound: 404 GET https://bigquery.googleapis.com/bigquery/v2/projects/bhearsum-buildhub-bigquery/datasets/buildhubtest/tables/builds2: Not found: Table bhearsum-buildhub-bigquery:buildhubtest.builds2
```

acmiyaguchi replied:
I can reproduce this in a test under `test_main_rebuild_bigquery`, which reveals that a missing dataset is also problematic.

```
google.api_core.exceptions.NotFound: 404 POST https://bigquery.googleapis.com/bigquery/v2/projects/amiyaguchi-dev/datasets/testing_not_exists_dataset/tables: Not found: Dataset amiyaguchi-dev:testing_not_exists_dataset
```

The check is working at expected though, since the the table and dataset should exist before insertions can be made. I can add dataset and table creation as part of app startup, though it would make `bigquery:admin` permissions a requirement for the service account.

mozbhearsum replied:
Hm, I thought we were going to require this anyways because of https://github.com/mozilla-releng/buildhub2/pull/617/files#diff-67995dccf929276c3452a5352d57e030R48 ?

Could we pre-provision them before we deploy this, and get rid of the drop/create?

acmiyaguchi replied:
That's true, though I'm not entirely sure what roles will be provisioned. If this gets placed under `moz-fx-data-shared-prod`, we will want to make sure that the service account is not able to modify anything outside of its own dataset.

> Could we pre-provision them before we deploy this, and get rid of the drop/create?

In this PR, no. The drop/create clause is there to ensure multiple runs of this command do not create duplicates. We can get rid of this once logic is implemented to only incrementally update the table.

mozbhearsum replied:
Just to be 100% crystal clear, we need credentials and a dataset preprovisioned, and those credentials need to be able to drop and create tables within that dataset?

acmiyaguchi replied:
Yes, and those credentials can be tested via the docker-compose tests here.

mozbhearsum commented on docker-compose.yml:
```diff
@@ -58,8 +58,11 @@ services:
       - DJANGO_ES_URLS=http://elasticsearch:9200
       - DATABASE_URL=postgres://buildhub2:buildhub2@db/buildhub2
       - DEVELOPMENT=1
+      - GOOGLE_APPLICATION_CREDENTIALS=/tmp/credentials
```
This is a nit pick, but I'd prefer this to be simply unset by default, and have the mount commented out with an explanation of how to enable it. Mounting arbitrary files isn't useful IMO.

acmiyaguchi replied:
I prefer the configuration set up this way, since development only requires `GOOGLE_APPLICATION_CREDENTIALS` to be set on the local machine for the rest of the machinery to work as intended. It's no longer mounting an arbitrary file (the comment is misleading since I modified the line in a previous commit). Just to clarify what this particular statement is doing, the docker-compose file is equivalent to the following command:

```
docker run  \
-v $GOOGLE_APPLICATION_CREDENTIALS:/tmp/credentials
-e GOOGLE_APPLICATION_CREDENTIALS=/tmp/credentials
-it buildhub2:latest
```

The environment variable `GOOGLE_APPLICATION_CREDENTIALS` points to the JSON keyfile on the local disk. This is mounted into the docker container if it exists. In the container, the application will look for a valid keyfile at the location `/tmp/credentials`.

acmiyaguchi replied:
Ah actually, I get the following error when unsetting `GOOGLE_APPLICATION_CREDENTIALS`, so maybe I'll add this to the README. `ERROR: Cannot create container for service web: create .: volume name is too short, names should be at least two alphanumeric characters`

mozbhearsum commented on tests/conftest.py:
```diff
@@ -95,6 +101,50 @@ def test_something(elasticsearch):
     build_index.delete(ignore=404)
 
 
+@runif_bigquery_testing_enabled
```
Does this need to be part of the fixture, when it's already applied to the consumers of the fixture?

acmiyaguchi replied:
I checked, and it doesn't need to be.

mozbhearsum commented on tests/test_main_models.py:
```diff
@@ -24,6 +27,111 @@ def test_insert(settings, valid_build):
+@pytest.mark.django_db
+@mock.patch("buildhub.main.bigquery.bigquery")
+def test_insert_skips_writes_to_bigquery_when_disabled(
+    mocked_bigquery, valid_build, settings
```
Thank you for improving the tests!

mozbhearsum commented on tests/utils.py:
```diff
@@ -0,0 +1,29 @@
+    bigquery.Client()
+    BIGQUERY_AUTHENTICATED = True
+except Exception:
+    # google.auth.exceptions.DefaultCredentialsError
```
It sounds like we know specifically what error we're expecting. If that's the case, let's catch just that. If others come up, it would be useful to have their tracebacks intsead of being eaten.

acmiyaguchi commented:
@mozbhearsum bumping for re-review. This should handle table creation during app startup. The dataset is required to exist before enabling BigQuery support.

acmiyaguchi commented:
I filed https://github.com/mozilla-releng/buildhub2/issues/618 to keep track of the issue. I can likely follow up sometime this month. I'm curious to see what the de-sync rate is in the first couple weeks of operation will be, too. Filling in the gaps will let us see what the actual rate of error.
